third/nmh/man/scan.man
.\"
.\" %nmhwarning%
.\" $Id: scan.man,v 1.1.1.1 1999-02-07 18:14:22 danw Exp $
.\"
.\" include the -mh macro file
.so %etcdir%/tmac.h
.\"
.TH SCAN %manext1% MH.6.8 [%nmhversion%]
.SH NAME
scan \- produce a one line per message scan listing
.SH SYNOPSIS
.in +.5i
.ti -.5i
scan
\%[+folder] \%[msgs]
\%[\-clear] \%[\-noclear]
\%[\-form\ formatfile]
\%[\-format\ string]
\%[\-header] \%[\-noheader]
\%[\-width\ columns]
\%[\-reverse] \%[\-noreverse]
\%[\-file filename]
.br
\%[\-version]
\%[\-help]
.in -.5i
.SH DESCRIPTION
\fIScan\fR produces a one\-line\-per\-message listing of the specified
folder or messages.  Each \fIscan\fR line contains the message number
(name), the date, the \*(lqFrom:\*(rq field, the \*(lqSubject\*(rq field,
and, if room allows, some of the body of the message.  For example:

.nf
.in +.5i
.ta \w'15+- 'u +\w'07/\|05x 'u +\w'Dcrocker  'u
15+	10/\|05 crocker	nned\0\0\*(<<Last week I asked some of
16\-	10/\|05 crocker	message id format\0\0\*(<<I recommend
18	10/\|06 brien	Re: Exit status from mkdir
19	10/\|07*brien	\*(lqscan\*(rq listing format in nmh
.re
.in -.5i
.fi

The `+' on message 15 indicates that it is the current message.

The `\-' on message 16 indicates that it has been replied to, as indicated
by a \*(lqReplied:\*(rq component (produced by the `\-annotate' switch
to the \fIrepl\fR command).

The `*' on message 19 indicates that no \*(lqDate:\*(rq header was
present.  The time of last modification of the message is given instead.

If there is sufficient room left on the \fIscan\fR line after the
subject, the line will be filled with text from the body, preceded by
<<, and terminated by >> if the body is sufficiently short.  \fIScan\fR
actually reads each of the specified messages and parses them to extract
the desired fields.  During parsing, appropriate error messages will be
produced if there are format errors in any of the messages.

By default, \fIscan\fR will decode RFC-2047 (MIME) encoding in
these scan listings.  \fIScan\fR will only decode these fields if your
terminal can natively display the character set used in the encoding.
You should set the MM_CHARSET environment variable to your native
character set, if it is not US-ASCII.  See the mh-profile(5) man
page for details about this environment variable.

The switch `\-reverse', makes \fIscan\fR list the messages in reverse
order.

The `\-file filename' switch allows the user to obtain a \fIscan\fP
listing of a maildrop file as produced by \fIpackf\fP.  This listing
includes every message in the file (you can't scan individual messages).
The switch `\-reverse' is ignored with this option.

The switch `\-width\ columns' may be used to specify the width of
the scan line.  The default is to use the width of the terminal.

The `\-header' switch produces a header line prior to the \fIscan\fR
listing.  Currently, the name of the folder and the current date and
time are output (see the \fBHISTORY\fR section for more information).

If the `\-clear' switch is used and \fIscan's\fR output is directed
to a terminal, then \fIscan\fR will consult the environment variables
\fB$TERM\fR and \fB$TERMCAP\fR to determine your terminal type in order
to find out how to clear the screen prior to exiting.  If the `\-clear'
switch is used and \fIscan's\fR output is not directed to a terminal
(e.g., a pipe or a file), then \fIscan\fR will send a formfeed prior
to exiting.

For example, the command:

.ti +.5i
(scan \-clear \-header; show all \-show pr \-f) | lpr

produces a scan listing of the current folder, followed by a formfeed,
followed by a formatted listing of all messages in the folder, one
per page.  Omitting `\-show\ pr\ \-f' will cause the messages to be
concatenated, separated by a one\-line header and two blank lines.

To override the output format used by \fIscan\fR, the `\-format\ string'
or `\-form\ file' switches are used.  This permits individual fields of
the scan listing to be extracted with ease.  The string is simply a format
string and the file is simply a format file.  See \fImh\-format\fR(5)
for the details.

In addition to the standard \fImh\-format\fR(5) escapes, \fIscan\fR
also recognizes the following additional \fIcomponent\fR escapes:
.sp 1
.nf
.ta \w'Dtimenow  'u +\w'Returns  'u
\fIEscape\fR	\fIReturns\fR	\fIDescription\fR
body	string	the (compressed) first part of the body
dtimenow	date	the current date
folder	string	the name of the current folder
.re
.fi

If no date header is present in the message, the \fIfunction\fR escapes
which operate on {\fIdate\fP\|} will return values for the date of last
modification of the message file itself.  This feature is handy for
scanning a \fIdraft folder\fR, as message drafts usually aren't allowed
to have dates in them.

\fIscan\fR will update the \fInmh\fR context prior to starting the listing,
so interrupting a long \fIscan\fR listing preserves the new context.
\fInmh\fR purists hate this idea.
.Fi
^$HOME/\&.mh\(ruprofile~^The user profile
.Pr
^Path:~^To determine the user's nmh directory
.Ps
^Alternate\-Mailboxes:~^To determine the user's mailboxes
.Ps
^Current\-Folder:~^To find the default current folder
.Sa
inc(1), pick(1), show(1), mh\-format(5)
.De
`+folder' defaults to the current folder
.Ds
`msgs' defaults to all
.Ds
`\-format' defaulted as described above
.Ds
`\-noheader'
.Ds
`\-width' defaulted to the width of the terminal
.Co
If a folder is given, it will become the current folder.
.Hi
Prior to using the format string mechanism, `\-header' used to generate
a heading saying what each column in the listing was.  Format strings
prevent this from happening.
.Bu
The argument to the `\-format' switch must be interpreted as a single
token by the shell that invokes \fIscan\fR.  Therefore, one must usually
place the argument to this switch inside double\-quotes.

The value of each \fIcomponent\fR escape is set by \fIscan\fR to the
contents of the first message header \fIscan\fR encounters with the
corresponding component name; any following headers with the same
component name are ignored.
.En


doc/news
This file is intended to keep track of user-visible changes to the
release, for the construction of the Athena release notes.

Changes from release 8.0 to release 8.1:

	* Athena version numbers are now dotted triplets, e.g. 8.1.0
	  or 8.1.12.  Paths which include all or part of the Athena
	  version number now take the form "pathname-version",
	  e.g. "srvd-8.1" or "dash.menus-8.1"

	* In release 8.0, running "xmkmf" with the default path would
	  generate a Makefile which looked for its include files and
	  libraries in /mit/x11, which could be one of several
	  lockers.  Now it is preferrable to use the vendor X
	  libraries, so this bit of glue is going away.  There will be
	  no /usr/athena/bin/xmkmf any more, nor any symlinks in
	  /usr/athena pointing to the x11 locker.

	  Users can still do something like "add -f x11r5" and then
	  "xmkmf", and things should work, if the x11r5 locker is
	  built for their platform.

	* Athena bitmaps have moved to /usr/athena/include/X11/bitmaps.
	  Previously, they were in /usr/athena/lib/X11/bitmaps before
	  because of the /mit/x11 glue.

	* The symlink /bin/csh -> /bin/athena/tcsh has gone away.  The
	  update to 8.0 will fix /bin/csh references in /etc/passwd
	  and /etc/passwd.local to refer to /bin/athena/tcsh instead.

	* emacs18 has gone away.

	* AFS includes and libraries are installed in
	  /usr/afsws/include and /usr/afsws/lib.

	* wcl includes and libraries are installed in /usr/athena.

	* The basic krb5 headers and libraries are installed in
	  /usr/athena.

	* emacs has been upgraded from 19.30 to 19.34.  There should
	  be no significant user-visible changes.

	* RCS has been upgraded from version 4 to version 5.7.

	* x3270 has been upgraded from version 1.2 to version 3.1.1.3.

	* sysinfo is now available under Irix.

	* /bin/athena/tcsh now has Hesiod support enabled, which means
	  that "~lockername" will work in a pathname in most cases.
	  (It won't work for users with local passwd information who
	  don't have their lockers attached.)

	* For improved ease of debugging, many binaries are installed
	  with symbol tables, and compiled without optimization.
	  Therefore, the contents of /bin/athena and /usr/athena/bin
	  may be somewhat larger, and the binaries may consume more
	  CPU cycles, although it's unlikely that anyone will notice
	  the speed difference.


third/nmh/uip/scan.c

/*
 * scan.c -- display a one-line "scan" listing of folder or messages
 *
 * $Id: scan.c,v 1.1.1.1 1999-02-07 18:14:16 danw Exp $
 */

#include <h/mh.h>
#include <h/fmt_scan.h>
#include <h/scansbr.h>
#include <zotnet/tws/tws.h>
#include <errno.h>

/*
 * We allocate space for message names (msgs array)
 * this number of elements at a time.
 */
#define MAXMSGS  256


static struct swit switches[] = {
#define	CLRSW	0
    { "clear", 0 },
#define	NCLRSW	1
    { "noclear", 0 },
#define	FORMSW	2
    { "form formatfile", 0 },
#define	FMTSW	3
    { "format string", 5 },
#define	HEADSW	4
    { "header", 0 },
#define	NHEADSW	5
    { "noheader", 0 },
#define	WIDTHSW	6
    { "width columns", 0 },
#define	REVSW	7
    { "reverse", 0 },
#define	NREVSW	8
    { "noreverse", 0 },
#define	FILESW	9
    { "file file", 4 },
#define VERSIONSW 10
    { "version", 0 },
#define	HELPSW	11
    { "help", 4 },
    { NULL, 0 }
};

extern int errno;

/*
 * global for sbr/formatsbr.c - yech!
 */
#ifdef LBL
extern struct msgs *fmt_current_folder;	
#endif

/*
 * prototypes
 */
void clear_screen(void);  /* from termsbr.c */


int
main (int argc, char **argv)
{
    int clearflag = 0, hdrflag = 0, ontty;
    int width = 0, revflag = 0;
    int i, state, msgnum, nummsgs, maxmsgs;
    int seqnum[NUMATTRS], unseen, num_unseen_seq = 0;
    char *cp, *maildir, *file = NULL, *folder = NULL;
    char *form = NULL, *format = NULL, buf[BUFSIZ];
    char **argp, *nfs, **arguments, **msgs;
    struct msgs *mp;
    FILE *in;

#ifdef LOCALE
    setlocale(LC_ALL, "");
#endif
    invo_name = r1bindex (argv[0], '/');

    /* read user profile/context */
    context_read();

    mts_init (invo_name);
    arguments = getarguments (invo_name, argc, argv, 1);
    argp = arguments;

    /*
     * Allocate the initial space to record message
     * names, ranges, and sequences.
     */
    nummsgs = 0;
    maxmsgs = MAXMSGS;
    if (!(msgs = (char **) malloc ((size_t) (maxmsgs * sizeof(*msgs)))))
	adios (NULL, "unable to allocate storage");

    /*
     * Parse arguments
     */
    while ((cp = *argp++)) {
	if (*cp == '-') {
	    switch (smatch (++cp, switches)) {
		case AMBIGSW: 
		    ambigsw (cp, switches);
		    done (1);
		case UNKWNSW: 
		    adios (NULL, "-%s unknown", cp);

		case HELPSW: 
		    snprintf (buf, sizeof(buf), "%s [+folder] [msgs] [switches]",
			invo_name);
		    print_help (buf, switches, 1);
		    done (1);
		case VERSIONSW:
		    print_version(invo_name);
		    done (1);

		case CLRSW: 
		    clearflag++;
		    continue;
		case NCLRSW: 
		    clearflag = 0;
		    continue;

		case FORMSW: 
		    if (!(form = *argp++) || *form == '-')
			adios (NULL, "missing argument to %s", argp[-2]);
		    format = NULL;
		    continue;
		case FMTSW: 
		    if (!(format = *argp++) || *format == '-')
			adios (NULL, "missing argument to %s", argp[-2]);
		    form = NULL;
		    continue;

		case HEADSW: 
		    hdrflag++;
		    continue;
		case NHEADSW: 
		    hdrflag = 0;
		    continue;

		case WIDTHSW: 
		    if (!(cp = *argp++) || *cp == '-')
			adios (NULL, "missing argument to %s", argp[-2]);
		    width = atoi (cp);
		    continue;
		case REVSW:
		    revflag++;
		    continue;
		case NREVSW:
		    revflag = 0;
		    continue;

		case FILESW:
		    if (!(cp = *argp++) || (cp[0] == '-' && cp[1]))
			adios (NULL, "missing argument to %s", argp[-2]);
		    if (strcmp (file = cp, "-"))
			file = path (cp, TFILE);
		    continue;
	    }
	}
	if (*cp == '+' || *cp == '@') {
	    if (folder)
		adios (NULL, "only one folder at a time!");
	    else
		folder = path (cp + 1, *cp == '+' ? TFOLDER : TSUBCWF);
	} else {
	    /*
	     * Check if we need to allocate more space
	     * for message names/ranges/sequences.
	     */
	    if (nummsgs >= maxmsgs) {
		maxmsgs += MAXMSGS;
		if (!(msgs = (char **) realloc (msgs,
			(size_t) (maxmsgs * sizeof(*msgs)))))
		    adios (NULL, "unable to reallocate msgs storage");
	    }
	    msgs[nummsgs++] = cp;
	}
    }

    if (!context_find ("path"))
	free (path ("./", TFOLDER));

    /*
     * Get new format string.  Must be before chdir().
     */
    nfs = new_fs (form, format, FORMAT);

    /*
     * We are scanning a maildrop file
     */
    if (file) {
	if (nummsgs)
	    adios (NULL, "\"msgs\" not allowed with -file");
	if (folder)
	    adios (NULL, "\"+folder\" not allowed with -file");

	/* check if "file" is really stdin */
	if (strcmp (file, "-") == 0) {
	    in = stdin;
	    file = "stdin";
	} else {
	    if ((in = fopen (file, "r")) == NULL)
		adios (file, "unable to open");
	}

#ifndef	JLR
	if (hdrflag) {
	    printf ("FOLDER %s\t%s\n", file, dtimenow (1));
	}
#endif /* JLR */

	m_unknown (in);
	for (msgnum = 1; ; ++msgnum) {
	    state = scan (in, msgnum, -1, nfs, width, 0, 0,
		    hdrflag ? file : NULL, 0L, 1);
	    if (state != SCNMSG && state != SCNENC)
		break;
	}
	fclose (in);
	done (0);
    }

    /*
     * We are scanning a folder
     */

    if (!nummsgs)
	msgs[nummsgs++] = "all";
    if (!folder)
	folder = getfolder (1);
    maildir = m_maildir (folder);

    if (chdir (maildir) == NOTOK)
	adios (maildir, "unable to change directory to");

    /* read folder and create message structure */
    if (!(mp = folder_read (folder)))
	adios (NULL, "unable to read folder %s", folder);

    /* check for empty folder */
    if (mp->nummsg == 0)
	adios (NULL, "no messages in %s", folder);

    /* parse all the message ranges/sequences and set SELECTED */
    for (msgnum = 0; msgnum < nummsgs; msgnum++)
	if (!m_convert (mp, msgs[msgnum]))
	    done(1);
    seq_setprev (mp);			/* set the Previous-Sequence */

    context_replace (pfolder, folder);	/* update current folder         */
    seq_save (mp);			/* synchronize message sequences */
    context_save ();			/* save the context file         */

    /*
     * Get the sequence number for each sequence
     * specified by Unseen-Sequence
     */
    if ((cp = context_find (usequence)) && *cp) {
	char **ap, *dp;

	dp = getcpy(cp);
	ap = brkstring (dp, " ", "\n");
	for (i = 0; ap && *ap; i++, ap++)
	    seqnum[i] = seq_getnum (mp, *ap);

	num_unseen_seq = i;
	if (dp)
	    free(dp);
    }

    ontty = isatty (fileno (stdout));

#ifdef LBL
    else
	fmt_current_folder = mp;
#endif

    for (msgnum = revflag ? mp->hghsel : mp->lowsel;
	 (revflag ? msgnum >= mp->lowsel : msgnum <= mp->hghsel);
	 msgnum += (revflag ? -1 : 1)) {
	if (is_selected(mp, msgnum)) {
	    if ((in = fopen (cp = m_name (msgnum), "r")) == NULL) {
#if 0
		if (errno != EACCES)
#endif
		    admonish (cp, "unable to open message");
#if 0
		else
		    printf ("%*d  unreadable\n", DMAXFOLDER, msgnum);
#endif
		continue;
	    }

#ifndef JLR
	    if (hdrflag) {
		printf ("FOLDER %s\t%s\n", folder, dtimenow(1));
	    }
#endif /* JLR */

	    /*
	     * Check if message is in any sequence given
	     * by Unseen-Sequence profile entry.
	     */
	    unseen = 0;
	    for (i = 0; i < num_unseen_seq; i++) {
		if (in_sequence(mp, seqnum[i], msgnum)) {
		    unseen = 1;
		    break;
		}
	    }

	    switch (state = scan (in, msgnum, 0, nfs, width,
			msgnum == mp->curmsg, unseen,
			hdrflag ? folder : NULL, 0L, 1)) {
		case SCNMSG: 
		case SCNENC: 
		case SCNERR: 
		    break;

		default: 
		    adios (NULL, "scan() botch (%d)", state);

		case SCNEOF: 
#if 0
		    printf ("%*d  empty\n", DMAXFOLDER, msgnum);
#else
		    advise (NULL, "message %d: empty", msgnum);
#endif
		    break;
	    }
	    hdrflag = 0;
	    fclose (in);
	    if (ontty)
		fflush (stdout);
	}
    }

#ifdef LBL
    seq_save (mp);	/* because formatsbr might have made changes */
#endif

    folder_free (mp);	/* free folder/message structure */
    if (clearflag)
	clear_screen ();

    done (0);
}
